world-clock/src/widgets/App.ts

```diff
@@ -0,0 +1,150 @@
+import on from '@dojo/core/on';
+import i18n, { switchLocale } from '@dojo/i18n/i18n';
+import { v, w } from '@dojo/widget-core/d';
+import I18nMixin from '@dojo/widget-core/mixins/I18n';
+import { theme, ThemedMixin } from '@dojo/widget-core/mixins/Themed';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+
+import * as moment from 'moment-timezone';
+
+import * as css from '../styles/app.m.css';
+import nlsBundle from '../nls/main';
+import Clock from './Clock';
+
+interface CityData {
+	key: string;
+	locale: string;
+	tz: string;
+}
+
+interface LanguageData {
+	key: string;
+	locale: string;
+	name: string;
+}
+
+export const AppBase = ThemedMixin(I18nMixin(WidgetBase));
+
+@theme(css)
+export default class App extends AppBase {
+	private _multipleLocales: boolean = false;
+	private _cities: CityData[] = [
+		{ key: 'mexicoCity', locale: 'es', tz: 'America/Mexico_City' },
+		{ key: 'newYork', locale: 'en', tz: 'America/New_York' },
+		{ key: 'london', locale: 'en', tz: 'Europe/London' },
+		{ key: 'johannesburg', locale: 'en', tz: 'Africa/Johannesburg' },
+		{ key: 'cairo', locale: 'ar', tz: 'Africa/Cairo' },
+		{ key: 'paris', locale: 'fr', tz: 'Europe/Paris' },
+		{ key: 'newDehli', locale: 'hi', tz: 'Asia/Colombo' },
+		{ key: 'tokyo', locale: 'ja', tz: 'Asia/Tokyo' }
+	];
+	private _languages: LanguageData[] = [
+		{ key: 'arabic', locale: 'ar', name: 'عربى' },
+		{ key: 'english', locale: 'en', name: 'English' },
+		{ key: 'spanish', locale: 'es', name: 'Español' },
+		{ key: 'french', locale: 'fr', name: 'Français' },
+		{ key: 'hindi', locale: 'hi', name: 'हिंदी' },
+		{ key: 'japanese', locale: 'ja', name: '日本語' }
+	];
+
+	constructor() {
+		super();
+		on(document, 'visibilitychange', () => {
```

Comment:
This feels like it's the perfect use-case for the `GlobalEvent` widget in `@dojo/widgets`, something like this:

```ts
private _onShow = () => {
	if (!document.hidden) {
		this.invalidate();
	}
}

protected render() {
	const { messages } = this.localizeBundle(nlsBundle);
	return w(GlobalEvent, { document: { visibilitychange: this._onShow } },  [
		v('div', { 
			lang: i18n.locale, 
			dir: i18n.locale.indexOf('ar-') === 0 ? 'rtl' : 'ltr'
		}, [
			this._renderFormInputs(messages),
			this._renderClocks()
		])
	]);
}
```

Reply:
Ah, that is perfect. I'll update that now.